Src/AutoFixture/DataAnnotations/AttributeRelay.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using Ploeh.AutoFixture.Kernel;
+
+namespace Ploeh.AutoFixture.DataAnnotations
+{
+    /// <summary>
+    /// Relays requests made for properties, fields or parameters to different requests, based on custom attributes 
+    /// applied to the code elements. 
+    /// </summary>
+    /// <remarks>
+    /// A typical example is relaying a request for a parameter that has <see cref="RangeAttribute"/> applied to it 
+    /// to a <see cref="RangedNumberRequest"/>. Another way to describe it, a request for a <see cref="ParameterInfo"/>
+    /// instance that contains a <see cref="RangeAttribute"/>, is transformed to a <see cref="RangedNumberRequest"/>
+    /// with parameters extracted from the attribute.
+    /// </remarks>
+    /// <typeparam name="TAttribute">
+    /// An <see cref="Attribute"/> type that can be applied to a property, a field or a method. The attribute is 
+    /// expected to allow applying only once. When the same attribute is applied multiple times to the same code 
+    /// element, an <see cref="InvalidOperationException"/> will be thrown.
+    /// </typeparam>
+    public class AttributeRelay<TAttribute> : ISpecimenBuilder

```

Comment:
My problem with having this class being a public type is that it introduces a new class (i.e. a new abstraction) that we'll have to maintain for at least two major versions of AutoFixture; that's a couple of years, at minimum. This abstraction is abstruse; while I understand what it does, it seems that its only purpose is to eliminate duplication.

Duplication is worth eliminating [if there's a cost associated with it](http://blog.ploeh.dk/2014/08/07/why-dry). In the case of `RangeAttributeRelay`, `RegularExpressionAttributeRelay`, and `StringLengthAttributeRelay`, these classes are extremely stable, so the cost of the duplication is low; the last time any of these classes were modified were in late 2012 (in a5b89aca2c9bb3ed0082a09ad26c227da9fc8175), and that was even in a part of its code not related to what's being extracted here.

These three classes are abstract enough as it is, but at least it seems fairly clear that they relay a data annotation to a 'proper' request object.

The proposed `AttributeRelay<TAttribute>` may benefit the developers of AutoFixture, but does it benefit the *users*?